client/exec.go

```diff
@@ -0,0 +1,305 @@
+}
+
+// ExecAdditionalArgs are additional control arguments for the Exec call.
+type ExecAdditionalArgs struct {
+	// Standard input, output, and error.
+	Stdin  io.ReadCloser
+	Stdout io.Writer
+	Stderr io.Writer
+
+	// Control message handler (for window resizing and signal forwarding)
+	Control func(conn WebsocketWriter)
+
+	// Channel that will be closed when all data operations are done
+	DataDone chan bool
+}
+
+// Exec starts a command execution with the given options and additional
+// control arguments, returning the execution's change ID.
+func (client *Client) Exec(opts *ExecOptions, args *ExecAdditionalArgs) (string, error) {
+	if opts.Terminal && opts.Stderr {
+		return "", errors.New("separate stderr not currently supported in terminal mode")
+	}
+	if !opts.Terminal && !opts.Stderr {
+		return "", errors.New("combined stderr not currently supported in non-terminal mode")
+	}
+
+	var timeoutStr string
+	if opts.Timeout != 0 {
+		timeoutStr = opts.Timeout.String()
+	}
+
+	var payload = struct {
+		Command     []string          `json:"command"`
+		Environment map[string]string `json:"environment"`
+		WorkingDir  string            `json:"working-dir"`
+		Timeout     string            `json:"timeout"`
+		UserID      *int              `json:"user-id"`
+		User        string            `json:"user"`
+		GroupID     *int              `json:"group-id"`
+		Group       string            `json:"group"`
+		Terminal    bool              `json:"terminal"`
+		Stderr      bool              `json:"stderr"`
+		Width       int               `json:"width"`
+		Height      int               `json:"height"`
+	}{
+		Command:     opts.Command,
+		Environment: opts.Environment,
+		WorkingDir:  opts.WorkingDir,
+		Timeout:     timeoutStr,
+		UserID:      opts.UserID,
+		User:        opts.User,
+		GroupID:     opts.GroupID,
+		Group:       opts.Group,
+		Terminal:    opts.Terminal,
+		Stderr:      opts.Stderr,
+		Width:       opts.Width,
+		Height:      opts.Height,
+	}
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(&payload)
+	if err != nil {
+		return "", err
+	}
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+	resultBytes, changeID, err := client.doAsyncFull("POST", "/v1/exec", nil, headers, &body)
+	if err != nil {
+		return "", err
+	}
+	var result struct {
+		WebsocketIDs map[string]string `json:"websocket-ids"`
+	}
+	err = json.Unmarshal(resultBytes, &result)
+	if err != nil {
+		return "", err
+	}
+
+	if args == nil {
+		return changeID, nil
+	}
+
+	// Process additional arguments (connecting I/O and websockets)
+	fds := result.WebsocketIDs
+
+	// Call the control handler with a connection to the control socket
+	if args.Control != nil && fds["control"] != "" {
+		conn, err := client.getChangeWebsocket(changeID, fds["control"])
+		if err != nil {
+			return "", err
+		}
+		go args.Control(conn)
+	}
+
+	if opts.Terminal {
+		// Handle terminal-based executions
+		if args.Stdin != nil && args.Stdout != nil {
```

Comment:
This forces the presence of Stdin to be able to read Stdout, but if we think about how we handle commands locally, stdin is almost always empty. I suggest mimicking that in the API as well, and accepting a nil Stdin to mean empty.

Reply:
Good point. I've simplified this considerably, and `Exec` now allows stdin to be nil (meaning no input) and stdout/stderr to be nil as well (meaning discard output).